keps/sig-scheduling/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,885 @@
+    // The name of each capacity must be unique in that set and
+    // include the domain prefix.
+    //
+    // Removing a capacity is not supported. It can be reduced to 0 instead.
+    //
+    // The maximum number of attributes and capacities in the DevicePatch combined is 32.
+    // This is an alpha field and requires enabling the DRAAdminControlledDeviceAttributes
+    // feature gate.
+    //
+    // +optional
+    // +featureGate:DRAAdminControlledDeviceAttributes
+    Capacity map[FullyQualifiedName]DeviceCapacity
+
+    // ^^^^
+    // The assumption here is that all device types will have attributes and capacities,
+    // similar to the current BasicDevice type. Therefore the overrides are not made
+    // specific to certain device types.
+}
+
+// DevicePatchFilter defines which device(s) a [DevicePatch] applies to.
+type DevicePatchFilter struct {
+    // If DeviceClassName is set, the selectors defined there must be
+    // satisfied by a device to be patched. This field corresponds
+    // to class.metadata.name.
+    //
+    // +optional
+    DeviceClassName *string
+
+    // If driver is set, only devices from that driver are patched.
+    // This fields corresponds to slice.spec.driver.
+    //
+    // +optional
+    Driver *string
+
+    // If pool is set, only devices in that pool are patched.
+    // This fields corresponds to slice.spec.pool.name.
+    //
+    // Also setting the driver name may be useful to avoid
+    // ambiguity when different drivers use the same pool name,
+    // but this is not required because selecting pools from
+    // different drivers may also be useful, for example when
+    // drivers with node-local devices use the node name as
+    // their pool name.
+    //
+    // +optional
+    Pool *string
+
+    // If device is set, only devices with that name are patched.
+    // This field corresponds to slice.spec.devices[].name.
+    //
+    // Also setting driver and pool may be required to avoid ambiguity,
+    // but is not required.
+    //
+    // +optional
+    Device *string
+
+    // Selectors define criteria which must be satisfied by a
+    // device to be patched. All selectors must be satisfied.
+    //
+    // +optional
+    // +listType=atomic
+    Selectors []DeviceSelector
+
+    // ^^^
+    //
+    // Selectors is a list for the same reason why a request has a list: at some point
+    // we might have one entry which uses CEL while another uses a different, yet to be
+    // defined mechanism.
+}
+```
+
+To distinguish intentionally empty attributes from attributes which have some
+future, unknown content and thus only seem to be empty to an older client, a
+special "null value" gets introduced:
+
+```golang
+// NullableDeviceAttribute must have exactly one field set.
+// It has the exact same fields as a [DeviceAttribute] plus NullValue as
+// an additional alternative.
+type NullableDeviceAttribute struct {
+    DeviceAttribute
```

Comment:
An observation, just in case this is unexpected: defining this with only a JSON struct tag like this and updating codegen results in a protobuf tag added unlike when `metav1.TypeMeta` is embedded.

Before:
```go
	DeviceAttribute `json:",inline"`
```
After:
```go
	DeviceAttribute `json:",inline" protobuf:"bytes,1,opt,name=deviceAttribute"`
```

The OpenAPI still looks right to me though in that it flattens all the value types into one map, but will it cause any problems for the generated protobuf to have that intermediate step in between?
```proto
message NullableDeviceAttribute {
  optional DeviceAttribute deviceAttribute = 1;

  // NullValue, if set, marks an intentionally empty attribute.
  //
  // +optional
  // +oneOf=ValueType
  optional NullValue null = 2;
}
```

Reply:
The KEP isn't trying to be a full spec for the API. I am intentionally leaving out most struct tags because they tend to be unnecessary and distract readers (see [comment earlier](https://github.com/kubernetes/enhancements/pull/5034#discussion_r1934109036) when I had copy-and-pasted them from Go code).

I think the protobuf is fine with that extra struct. I was expecting it to be like this when I asked about whether we would need a real, flat struct for declarative validation: that would be a breaking change for the protobuf encoding.